src/instruction_graph_generator.cc

```diff
@@ -0,0 +1,2038 @@
+		}
+	} else {
+		for(min[dim] = full_box_min[dim]; min[dim] < full_box_max[dim]; min[dim] += communicator_max_coordinate) {
+			max[dim] = std::min(full_box_max[dim], min[dim] + communicator_max_coordinate);
+			comm_boxes.emplace_back(min, max);
+		}
+	}
+}
+
+/// The (MPI) communicator can only exchange boxes sized up to `communicator_max_coordinate` in each dimension. Larger boxes must be split - at the implied
+/// transfer sizes, this will not have any measurable performance impact.
+box_vector<3> split_into_communicator_compatible_boxes(const range<3>& buffer_range, const box<3>& full_box) {
+	assert(box(subrange<3>(zeros, buffer_range)).covers(full_box));
+
+	int slice_dim = 0;
+	for(int d = 1; d < 3; ++d) {
+		if(buffer_range[d] > communicator_max_coordinate) { slice_dim = d; }
+	}
+
+	box_vector<3> comm_boxes;
+	split_into_communicator_compatible_boxes_recurse(comm_boxes, full_box, full_box.get_min(), full_box.get_max(), slice_dim, 0);
+	return comm_boxes;
+}
+
+/// Determines whether two boxes are either overlapping or touching on edges (not corners). This means on 2-connectivity for 1d boxes, 4-connectivity for 2d
+/// boxes and 6-connectivity for 3d boxes. For 0-dimensional boxes, always returns true.
+template <int Dims>
+bool boxes_edge_connected(const box<Dims>& box1, const box<Dims>& box2) {
+	if(box1.empty() || box2.empty()) return false;
+
+	// boxes can be 2/4/6 connected by either fully overlapping, or by overlapping in Dims-1 dimensions and touching (a.max[d] == b.min[d]) in the remaining one
+	bool disconnected = false;
+	int n_dims_touching = 0;
+	for(int d = 0; d < Dims; ++d) {
+		// compute the intersection but without normalizing the box to distinguish the "disconnected" from the "touching" case
+		const auto min = std::max(box1.get_min()[d], box2.get_min()[d]);
+		const auto max = std::min(box1.get_max()[d], box2.get_max()[d]);
+		if(min < max) {
+			// boxes overlap in this dimension
+		} else if(min == max) {
+			n_dims_touching += 1;
+		} else /* min > max */ {
+			disconnected = true;
+		}
+	}
+	return !disconnected && n_dims_touching <= 1;
+}
+
+// explicit instantiations for tests
+template bool boxes_edge_connected(const box<0>& box1, const box<0>& box2);
+template bool boxes_edge_connected(const box<1>& box1, const box<1>& box2);
+template bool boxes_edge_connected(const box<2>& box1, const box<2>& box2);
+template bool boxes_edge_connected(const box<3>& box1, const box<3>& box2);
+
+/// Subdivide a region into connected partitions (where connectivity is established by `boxes_edge_connected`) and return the bounding box of each partition.
+/// Note that the returned boxes are not necessarily disjoint event through the partitions always are.
+///
+/// This logic is employed to find connected subregions in a pending-receive that might be satisfied by a peer through a single send operation and thus requires
+/// a contiguous backing allocation.
```

Comment:
It can happen if the host allocation triggers any resize-copies, which are split on the producer and not the consumer side. Typically, these allocations will happen once without any dependencies of their own.

Receives that have multiple consumer-instructions will be initiated as a `split_receive` (on the connected-subregion allocation as discussed), but awaited via `await_receives`, whose regions are determined by `symmetrically_split_overlapping_regions`.